// @sap/approuter/lib/middleware/logout-provider.js
'use strict';

var URI = require('urijs');
var urlUtils = require('../utils/url-utils');
var loggerUtils = require('../utils/logger');
var backendRequestOptions = require('../backend-request/options');
var uaaUtils = require('../utils/uaa-utils');

module.exports = {
  sessionTimeoutLogout: function(session, app) {
    if (session.user) {
      logout(app, null, session);
    }
  },

  centralLogout: function(req, res) {
    logout(req.app, req);

    req.session.destroy();
    res.removeHeader('set-cookie');
    module.exports.triggerUAARedirect(req, res, function(err) {
      if (err) { throw err; }
    });
  },

  callBackendLogoutPaths: function(req, routerConfig, accessToken) {

    var destinations = routerConfig.appConfig.destinations || {};

    for (var destinationName in destinations) {
      var destinationOptions = destinations[destinationName];
      if (destinationOptions.logoutPath) {
        var destination = routerConfig.destinations[destinationName];
        var logoutUri = urlUtils.join(destination.url, destinationOptions.logoutPath);
        var request = backendRequestOptions.getLogoutRequest(req, accessToken, destination, destinationOptions);
        sendRequest(request, destination, logoutUri);
      }
    }
  },

  triggerUAARedirect: function(req, res, cb) {
    resolveLogoutRedirectUrl(req, function (err, redirectUrl) {
      if (!err) {
        res.writeHead(302, { Location: redirectUrl });
        res.end();
      }
      cb(err);
    });
  }
};

function logout(app, req, session) {
  session = session || req.session;
  module.exports.callBackendLogoutPaths(req, app.get('mainRouterConfig'), session.user.token.accessToken);
  app.approuter.emit('logout', session);
}

function sendRequest(request, destination, logoutUri) {
  var logger = loggerUtils.getLogger('/Middleware');
  var tracer = loggerUtils.getTracer(__filename);

  var clientReq = request();
  if (destination.timeout) {
    clientReq.setTimeout(destination.timeout);
  }
  clientReq.on('timeout', function () {
    clientReq.abort();
    logger.error('Request to %s failed with a timeout', logoutUri);
  });
  clientReq.on('error', function (err) {
    logger.error(err, 'Error in request to %s', logoutUri);
  });
  clientReq.on('response', function (resp) {
    if (destination.timeout) {
      resp.setTimeout(destination.timeout);
    }
    resp.on('timeout', function () {
      clientReq.abort();
      logger.error('Request to %s failed with a timeout', logoutUri);
    });
    resp.on('error', function (err) {
      logger.error(err, 'Error in response from %s', logoutUri);
    });
    var body = '';
    resp.on('data', function (chunk) {
      body += chunk;
    });
    resp.on('end', function () {
      if (resp.statusCode !== 200) {
        logger.error('Status %d received for request to %s, body: %s', resp.statusCode, logoutUri, body);
      }
      tracer.info('Status %d received for request to %s, body: %s', resp.statusCode, logoutUri, body);
    });
  });
  clientReq.end();
}

function resolveLogoutRedirectUrl(req, cb) {
  // Currently redirect from UAA to application is possible only with port-based routing
  var logout = req.routerConfig.appConfig.logout;
  var logoutPage = logout && logout.logoutPage;

  uaaUtils.getUaaConfig(req, function (err, uaaConfig) {
    if (err) { return cb(err); }

    var redirectUrl = urlUtils.join(uaaConfig.url, '/logout.do');
    if (!logoutPage) {
      return cb(null, redirectUrl);
    }
    var uri = URI.parse(logoutPage);
    if (uri.protocol && uri.hostname) {
      return cb(null, redirectUrl + '?redirect=' + logoutPage);
    }
    var appRouterUrl = urlUtils.buildAppRouterUrl(req);
    cb(null, redirectUrl + '?redirect=' + urlUtils.join(appRouterUrl, logoutPage));
  });
}
