// @sap/grunt-sapui5module-bestpractice-build/tasks/index.js
'use strict';
var path = require('path');
module.exports = function(grunt) {
    var distDir = path.join('dist');

    grunt.initConfig({
        "codevalidation": {
            "di.code-validation.js": {
                options: {
                    projectPath: process.cwd(),
                    ignoredPaths: ['node_modules', distDir],
                    reporter: "problems_reporter",
                    reporterOptions: {outputFile: path.join(distDir, "di.code-validation.core_issues.json")},
                    validators: {
                        "@sap/di.code-validation.js": {
                            extensions: [".js"]
                        }
                    }
                }
            },
            "di.code-validation.xml": {
                options: {
                    projectPath: process.cwd(),
                    ignoredPaths: ['node_modules', distDir],
                    reporter: "problems_reporter",
                    reporterOptions: {outputFile: path.join(distDir, "di.code-validation.core_issues.json")},
                    validators: {
                        "@sap/di.code-validation.xml": {
                            extensions: [".xml"]
                        }
                    }
                }
            }

        }}
    );

    grunt.registerTask('lint', 'lint project', function () {
        grunt.loadNpmTasks('@sap/di.code-validation.grunt');
        grunt.loadNpmTasks('grunt-mkdir');

        grunt.config("mkdir", {
            distDir: {
                options: {
                    create: [distDir]
                }
            }
        });

        grunt.registerTask('di.code-validation.js', function () {
            grunt.task.run(['codevalidation:di.code-validation.js']);
        });

        grunt.registerTask('di.code-validation.xml', function () {
            grunt.task.run(['codevalidation:di.code-validation.xml']);

        });

        grunt.task.run([
            'mkdir:distDir',
            'di.code-validation.js',
            'di.code-validation.xml'
        ]);
    });	// Linting task


};
